### PR TITLE
ocamllex: Exact check for the absence of memory instructions.

### PR DESCRIPTION
**Pull request by maranget**
The PR corrects MPR 7760: Segfault in ocamllex-generated code using 'shortest'.

### CODE REVIEW

File: Changes
Comment (gasche) [inline Changes]: The reporter is Stephen Dolan (see [MPR#7760](https://caml.inria.fr/mantis/view.php?id=7760)). If you consider that the analysis work that Frédéric Bour ("def") did help you write your fix, you can credit him with "Luc Maranget, Frédéric Bour, report by Stephen Dolan".

Comment (gasche) [reply Changes]: (Also, the change entry could have the GPR number as well: `MPR#7760, GPR#1713`.)

Comment (gasche) [reply Changes]: Also, you chose the wrong category for Changes. You are in the "4.06 maintenance branch" which should in fact be marked "4.06.1" (I'll fix this), you should go in the "OCaml 4.07" part, just below (by numeric MPR ordering):

```
- MPR#7751, GPR#1657: The toplevel prints some concrete types as abstract
  (Jacques Garrigue, report by Matej Kosik)
```

**Review by gasche — COMMENTED** [review]
I did a non-expert review of the code (I had never looked at it before), and the patch makes sense from the explanation that Luc gave. I had two questions while reading the code:

1. Why is only `output` changed when `outputbis` contains the same code? (`output` generates C code, `outputbis` generates ML code, as the names suggest :-) The answer is that only `output`  uses compaction (why?), as can be seen in `main.ml`.

2. This PR disables emission of memory instruction when an action was using memory, but the compaction pass completely removed all memory actions. Wouldn't there be an intermediate case where some memory actions remain, but the ones for the state that we are processing were removed? Is there a risk of bug in this case?

File: Changes
Comment (maranget) [reply Changes]: Ok I change Change again.

**Comment by gasche** [comment]
I removed the suggestion to change `Outputbis` as I found it unconvincing when looking at the code again.

**Review by gasche — APPROVED** [review]
Approved. I would still like to check that the CI passes before this PR is merged.

File: Changes
Comment (gasche) [inline Changes]: I guess I'm `?`.

**Comment by maranget** [comment]
> gasche commented on this pull request.
> 
> 
> 
> > @@ -423,6 +423,10 @@ Working version
>  - MPR#7751, GPR#1657: The toplevel prints some concrete types as abstract
>    (Jacques Garrigue, report by Matej Kosik)
>  
> +- PR#7760, GPR#1713: Exact selection of lexing engine, that is
> +  correct "Segfault in ocamllex-generated code using 'shortest'"
> +  (Luc Maranget, Frédéric Bour, report by Stephen Dolan, Review by ?)
> 
> I guess I'm `?`.

You guess right, do I correct and merge ?

**Comment by gasche** [comment]
For now the CI results are not in: "Some checks haven't completed yet". You can either wait for them to finish, and then correct and merge (you can squash all your commits together with `git rebase -i trunk` for example), or you update the branch, and then we wait for the restarted CI to complete again -- I can take care of the merge.

**Pull request merged**
